File: 598/C-nearest-vectors/solution.py
import math

n = int(input())

vectors = []
for i in range(n):
    x, y = map(int, input().split())
    vectors.append((math.atan2(y, x), i + 1))
vectors.sort()

best = 2 * math.pi - abs(vectors[0][0]) - abs(vectors[n-1][0])
p1, p2 = vectors[0][1], vectors[n-1][1]
for i in range(n - 1):
    current = vectors[i+1][0] - vectors[i][0]
    print(current, vectors[i][1], vectors[i+1][1])
    if (current < best and current != 0):
        best = current
        p1, p2 = vectors[i][1], vectors[i+1][1]

print(p1, p2)

File: 598/C-nearest-vectors/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

const long double PI = atan(1) * 4;

struct PointVector {
    long double angle;
    int i;
};

bool cmp(PointVector a, PointVector b) {
    return a.angle < b.angle;
}

int main()
{
    int n, a, b, x, y;
    long double bestAns, currentAns;
    cin >> n;
    vector<PointVector> vectors(n);
    for (int i = 0; i < n; i++) {
        vectors[i].i = i + 1;
        cin >> x >> y;
        vectors[i].angle = atan2(y, x);
    }
    sort(vectors.begin(), vectors.end(), cmp);
    
    bestAns = (2 * PI - abs(vectors[0].angle) - abs(vectors[n - 1].angle));
    a = vectors[0].i; b = vectors[n-1].i;
    for (int i = 0; i < n - 1; i++) {
        currentAns = vectors[i + 1].angle - vectors[i].angle;
        if (currentAns < bestAns) {
            bestAns = currentAns;
            a = vectors[i].i;
            b = vectors[i + 1].i;
        }
    }
    cout << a << " " << b << endl;

    return 0;
}